### PR TITLE
[DT] Unify encoding materialization pass into a single pass.

### PR DESCRIPTION
**Pull request by hanhanW**
The revision creates a generic materialization pass and uses it for backends that implement data-tiling. After months of development, we identify that the needs of GPU is a superset of the needs of CPU. To be more specific, it has the additional "swizzle" field in terms of layout. It means that the GPU set_encoding/unset_encoding lowering patterns cover the needs of CPU path. The lowering of contraction ops is different. CPU lowers it to mmt4d ops, while GPU lowers it to multi_mma op. However, the lowering of contraction is implemented through attribute interface. Thus, we can have a generic pattern to lower contraction ops.

To make the review process much easier, the revision is created by 5 commits.

1. It directly creates the MaterializeEncoding pass and copy-paste the GPU patterns: SetEncodingOpLoweringConversion, UnSetEncodingOpLoweringConversion, and MaterializeContractionOp. In the first commit, it also updates the GPU tests to use the new pass.
2. The GPU data-tiling does not support element-wise generic op lowering atm. The second commit moves the pattern to shared pattern set and bail out when swizzle is present. This is an NFC for both pipelines.
3. The third commit replaces the existing materialization pass with the generic pass, and deletes all the legacy passes.
4. The four commit moves the lit tests from `Common/[CPU|GPU]/test` to `Common/test`.
5. Now there are duplicate patterns for set_encoding, unset_encoding, and contraction ops lowering. The last commit deletes the legacy patterns, and move the patterns from MaterializeEncoding.cpp to where the legacy patterns locate. Furthermore, it renames the file as `MaterializeEncodingPatterns.cpp`.

The revision retains the MaterializeEncodingIntoNop pass, and add a TODO item. Because it is still used by MaterializeHomogeneousEncoding pass. It can be deleted once we deprecate the early materialization path.

### CODE REVIEW

**Comment by hanhanW** [comment]
This depends on https://github.com/iree-org/iree/pull/19452 and https://github.com/iree-org/iree/pull/19453. I push the parent commits to an upstream branch, so it is ready for review.

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 1]
```diff
@@ -161,6 +161,11 @@ const char *getIreeArchNameForTargetTriple(llvm::Triple triple) {
   return "unknown";
 }
 
+bool isLLVMCPUBackend(IREE::HAL::ExecutableTargetAttr targetAttr) {
+  return targetAttr &&
+         targetAttr.getBackend().getValue().starts_with("llvm-cpu");
```
Comment (bjacob) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]: Can't you just check equality ? Why starts_with? If that was copying the vmvx case below, I might have done that out of ignorance in the past.

**Review by bjacob — APPROVED** [review]
Very nice reorganization!

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]: Oh, yes. Thanks for pointing it out! I'm mainly following the other cases, but can definitely check equality.

**Pull request merged**